web/BridgeAPI.ts

```diff
@@ -1,14 +1,72 @@
-import { BridgeRoomState } from '../src/Widgets/BridgeWidgetInterface';
-
+import { BridgeRoomState, GetConnectionsForServiceResponse, WidgetConfigurationSection } from '../src/Widgets/BridgeWidgetInterface';
+import { GetConnectionsResponseItem } from "../src/provisioning/api";
+import { ExchangeOpenAPIRequestBody, ExchangeOpenAPIResponseBody } from "matrix-appservice-bridge";
+import { WidgetApi } from 'matrix-widget-api';
+import { ApiError } from '../src/api';
 export class BridgeAPIError extends Error {
-    constructor(msg: string, private body: Record<string, unknown>) {
+    constructor(msg: string, public readonly body: ApiError) {
         super(msg);
     }
+
+    public get errcode() {
+        return this.body.errcode as string;
+    }
+    public get error() {
+        return this.body.error as string;
+    }
 }
 
 export default class BridgeAPI {
+    static async getBridgeAPI(baseUrl: string, widgetApi: WidgetApi): Promise<BridgeAPI> {
+        const sessionToken = localStorage.getItem('hookshot-sessionToken');
+        baseUrl = baseUrl.endsWith("/") ? baseUrl.slice(0, -1) : baseUrl;
+        if (sessionToken) {
+            const client = new BridgeAPI(baseUrl, sessionToken);
+            try {
+                await client.verify();
+                return client;
+            } catch (ex) {
+                // Clear the token from the server, also actually check the error here.
+                console.warn(`Failed to verify token, fetching new token`, ex);
+                localStorage.removeItem(sessionToken);
+            }
+        }
+        const creds = await widgetApi.requestOpenIDConnectToken();
+        const { matrix_server_name, access_token } = creds;
+        // eslint-disable-next-line camelcase
+        if (!matrix_server_name || !access_token) {
+            throw Error('Server OpenID response missing values');
+        }
 
-    constructor(private baseUrl: string, private roomId: string, private accessToken: string) {
+        const req = await fetch(`${baseUrl}/widgetapi/v1/exchange_openid`, {
```

Comment:
That should be `res`, right? It's a bit confusing now :)

Reply:
Yep, another good point 👍